### ISSUE TITLE
Same Autoencoder scores on different datasets

### ISSUE THREAD

**Issue by Arxa** [deeplearning4j/deeplearning4j#4452]
#### Issue Description

My goal is to train an Autoencoder on images containing text. Then, test the model against random images and see how probable is they contain text or not. 
**Problem is I am getting the same score 0.27 at all test images.**

**The Autoencoder** 
```
MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(12345)
                .iterations(10)
                .weightInit(WeightInit.XAVIER)
                .updater(Updater.ADAGRAD)
                .activation(Activation.RELU)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .learningRate(0.05)
                .regularization(true).l2(0.0001)
                .list()
                .layer(0, new DenseLayer.Builder().nIn(5000).nOut(1500)
                        .build())
                .layer(1, new DenseLayer.Builder().nIn(1500).nOut(250)
                        .build())
                .layer(2, new DenseLayer.Builder().nIn(250).nOut(10)
                        .build())
                .layer(3, new DenseLayer.Builder().nIn(10).nOut(250)
                        .build())
                .layer(4, new DenseLayer.Builder().nIn(250).nOut(1500)
                        .build())
                .layer(5, new OutputLayer.Builder().nIn(1500).nOut(5000)
                        .lossFunction(LossFunctions.LossFunction.MSE)
                        .build())
                .pretrain(true).backprop(true)
                .setInputType(InputType.convolutional(50,100,1))
                .build();

        MultiLayerNetwork net = new MultiLayerNetwork(conf);
```
**Training/Testing Data**
```
private static DataSetIterator getData(String type) throws IOException, InterruptedException {
        int height = 50;
        int width = 100;
        int channels = 1;
        int seed = 123;
        Random randNumGen = new Random(seed);
        int batchSize = 1;
        int outputNum = 5000;
        File modelData = new File("C:\\Users\\310297685\\Desktop\\"+type);
        FileSplit data = new FileSplit(modelData, NativeImageLoader.ALLOWED_FORMATS,randNumGen);
        ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();
        ImageRecordReader recordReader = new ImageRecordReader(height,width,channels,labelMaker);
        recordReader.initialize(data);
        recordReader.setListeners(new LogRecordListener());
        DataSetIterator dataIter = new RecordReaderDataSetIterator(recordReader,batchSize,1,outputNum);
        DataNormalization scaler = new ImagePreProcessingScaler(0,1);
        scaler.fit(dataIter);
        dataIter.setPreProcessor(scaler);
        return dataIter;
    }
```
**Directory Structure**
All images are grey scale with 50px height and 100px width. Number of training images are 392,240 while Testing1 and Testing2 images are 29 each.

![testing1_img](https://user-images.githubusercontent.com/15330998/34411449-97f1ece0-ebce-11e7-97a8-02803668c209.jpg)
![testing2_img](https://user-images.githubusercontent.com/15330998/34411450-98098332-ebce-11e7-89df-edee1f49446f.jpg)
![training_img](https://user-images.githubusercontent.com/15330998/34411451-981f77dc-ebce-11e7-99ca-2104e9f03434.jpg)

#### Version Information

```
buildscript {
    ext {
        dl4j_version = "0.9.1"
        kotlin_version = "1.1.4-3"
    }
    repositories {
        maven { url "https://oss.sonatype.org/content/repositories/snapshots" }
        mavenCentral()
    }
    dependencies {
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:${kotlin_version}"
    }
}

dependencies {
    ....
    ....
    compile "org.deeplearning4j:deeplearning4j-core:${dl4j_version}"
    compile "org.nd4j:nd4j-native-platform:${dl4j_version}"
}
```

**Training and saving the model**

```
DataSetIterator trainIter = getData("training");
int nEpochs = 30;
for (int epoch = 0; epoch < nEpochs; epoch++) {
   while (trainIter.hasNext()) {
        DataSet data = trainIter.next();
         net.fit(data.getFeatureMatrix());
   }
   System.out.println("Epoch " + epoch + " complete");
}
ModelSerializer.writeModel(net, new File("autoencoder.zip"), true);
```

**Loading and testing the model**

```
MultiLayerNetwork restored = ModelSerializer.restoreMultiLayerNetwork(new File("autoencoder.zip"));
DataSetIterator testIter = getData("testing1"); // tried 'testing2' -> same score results
while (testIter.hasNext()) {
      DataSet data = testIter.next();
      //INDArray scores = restored.scoreExamples(data,true); Tried this method also
      double score = restored.score(new DataSet(data.getFeatureMatrix(),data.getLabels()));
      System.out.println(score);
      //for (int i=0; i<scores.rows(); i++){
            //System.out.println("Score for "+i+" : "+scores.getDouble(i));
       //}
}
```

**Results - 'Testing1' directory**

```
[main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
[main] INFO org.reflections.Reflections - Reflections took 58 ms to scan 14 urls, producing 9199 keys and 9496 values 
[main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for NativeOps: 4
[main] INFO org.reflections.Reflections - Reflections took 210 ms to scan 13 urls, producing 31 keys and 227 values 
[main] INFO org.nd4j.nativeblas.Nd4jBlas - Number of threads used for BLAS: 4
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Windows 10]
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [8]; Memory: [3.5GB];
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [MKL]
[main] INFO org.reflections.Reflections - Reflections took 3 ms to scan 2 urls, producing 10 keys and 14 values 
[main] INFO org.deeplearning4j.nn.multilayer.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: NONE; inference: SEPARATE]
[main] INFO org.reflections.Reflections - Reflections took 147 ms to scan 13 urls, producing 387 keys and 1538 values 
[main] INFO Processors.Autoencoder - ***TESTING***
[main] INFO org.datavec.api.records.listener.impl.LogRecordListener - Reading C:\Users\310297685\Desktop\testing1\text\1_BOATING_8400.jpg.png
0.27558434708925855
0.2755859513601936
0.27566514631125105
0.2756167952141441
0.2755651857695259
0.2755566833576835
0.2756576011976875
0.27556430624338757
0.27562633481833115
0.27555694657179486
0.2755680895170845
0.2756095047077812
0.2755687565169014
0.2756204480728782
0.2755961991390861
0.27563872819754254
0.27558334742399826
0.27556892572256697
0.27563237048002853
0.2756046177229561
0.27577948098990096
0.27561357606264725
0.27556882265421523
0.27556978441092145
0.2755568767151512
0.27555836585375443
0.2755805631241478
0.2755597603401817
0.2755692813715614

Process finished with exit code 0
```

**Results - 'Testing2' directory**

```
[main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
[main] INFO org.reflections.Reflections - Reflections took 69 ms to scan 14 urls, producing 9199 keys and 9496 values 
[main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for NativeOps: 4
[main] INFO org.reflections.Reflections - Reflections took 232 ms to scan 13 urls, producing 31 keys and 227 values 
[main] INFO org.nd4j.nativeblas.Nd4jBlas - Number of threads used for BLAS: 4
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Windows 10]
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [8]; Memory: [3.5GB];
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [MKL]
[main] INFO org.reflections.Reflections - Reflections took 6 ms to scan 2 urls, producing 10 keys and 14 values 
[main] INFO org.deeplearning4j.nn.multilayer.MultiLayerNetwork - Starting MultiLayerNetwork with WorkspaceModes set to [training: NONE; inference: SEPARATE]
[main] INFO org.reflections.Reflections - Reflections took 188 ms to scan 13 urls, producing 387 keys and 1538 values 
[main] INFO Processors.MNISTAnomalyExample - ***TESTING***
[main] INFO org.datavec.api.records.listener.impl.LogRecordListener - Reading C:\Users\310297685\Desktop\testing2\text\non-text_000002.jpg
0.2756120784363426
0.27566625910612713
0.275568962129084
0.27562079847666393
0.27563516164633406
0.27555958090635907
0.2756057927212395
0.2756051555237449
0.27559729498240126
0.275582914384333
0.2757443964562095
0.275639523466555
0.2756213976225532
0.27562533806177747
0.2756052849850334
0.27555857842775955
0.27557888496729505
0.2756030116638817
0.27569099303099287
0.27566616164061203
0.275585767730204
0.27559361272665633
0.27564091509195937
0.275594069512812
0.2756252595982231
0.2756433569512047
0.275618678053347
0.27564530666681897
0.27562302642676006

Process finished with exit code 0
```
I am really new to DeepLearning4J and to Machine Learning in general. I apologise before hand in case I have made some stupid mistake/configuration (highly probable). Any help/insight appreciated. Cheers.

**Comment by AlexDBlack** [comment]
This could simply be a tuning issue - i.e., incorrect/unsuitable learning rate.

What I can suggest, however:
* Switch output layer activation function to sigmoid
* The bottleneck (10 units) may be too severe. Try making it a bit larger.
* ADAM updater is usually better than AdaGrad
* Use .iterations(1)

Use the following, instead of the doubly nested loops:
```
for (int epoch = 0; epoch < nEpochs; epoch++) {
   net.fit(trainIter);
}
```

Beyond that: post UI screenshots as per: https://deeplearning4j.org/visualization
Also perhaps have a look at: https://deeplearning4j.org/troubleshootingneuralnets

**Comment by Arxa** [comment]
I applied your suggestions but still getting same scores. 
Here is my updated code:
```
public static void main(String[] args) throws Exception {

        //Initialize the user interface backend
        UIServer uiServer = UIServer.getInstance();

        //Configure where the network information (gradients, score vs. time etc) is to be stored. Here: store in memory.
        StatsStorage statsStorage = new InMemoryStatsStorage();         //Alternative: new FileStatsStorage(File), for saving and loading later

        //Attach the StatsStorage instance to the UI: this allows the contents of the StatsStorage to be visualized
        uiServer.attach(statsStorage);

        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(12345)
                .iterations(1)
                .weightInit(WeightInit.XAVIER)
                .updater(Updater.ADAM)
                .activation(Activation.RELU)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .learningRate(0.05)
                .regularization(true).l2(0.0001)
                .list()
                .layer(0, new DenseLayer.Builder().nIn(5000).nOut(1500)
                        .build())
                .layer(1, new DenseLayer.Builder().nIn(1500).nOut(250)
                        .build())
                .layer(2, new DenseLayer.Builder().nIn(250).nOut(50)
                        .build())
                .layer(3, new DenseLayer.Builder().nIn(50).nOut(250)
                        .build())
                .layer(4, new DenseLayer.Builder().nIn(250).nOut(1500)
                        .build())
                .layer(5, new OutputLayer.Builder().nIn(1500).nOut(5000)
                        .lossFunction(LossFunctions.LossFunction.MSE)
                        .build())
                .pretrain(true).backprop(true)
                .setInputType(InputType.convolutional(50,100,1))
                .build();

        MultiLayerNetwork net = new MultiLayerNetwork(conf);
        net.setListeners(new ScoreIterationListener(1/5));

        //Then add the StatsListener to collect this information from the network, as it trains
        net.setListeners(new StatsListener(statsStorage));

        log.info("***TRAINING***");
        DataSetIterator trainIter = getData("training1");
        int nEpochs = 20;
        for (int epoch = 0; epoch < nEpochs; epoch++) {
            net.fit(trainIter);
            log.info("EPOCH: "+epoch);
        }

        log.info("Writing model");
        ModelSerializer.writeModel(net, new File("C:\\Users\\310297685\\IdeaProjects\\Thesis\\VideoText_Extractor\\src\\main\\resources\\models\\autoencoder.zip"), true);
        log.info("Writing done");
//        MultiLayerNetwork restored = ModelSerializer.restoreMultiLayerNetwork(new File("C:\\Users\\310297685\\IdeaProjects\\Thesis\\VideoText_Extractor\\src\\main\\resources\\models\\autoencoder.zip"));
//        log.info("***TESTING***");
//        DataSetIterator testIter = getData("testing2");
//        while (testIter.hasNext()) {
//            DataSet data = testIter.next();
//            //INDArray scores = restored.scoreExamples(data,true);
//            double score = restored.score(new DataSet(data.getFeatureMatrix(),data.getLabels()));
//            log.info(String.valueOf(score));
//        }
    }

private static DataSetIterator getData(String type) throws IOException, InterruptedException {
        int height = 50;
        int width = 100;
        int channels = 1;
        int seed = 123;
        Random randNumGen = new Random(seed);
        int batchSize = 100;
        int outputNum = 5000;
        File modelData = new File("C:\\Users\\310297685\\Desktop\\"+type);
        FileSplit data = new FileSplit(modelData, NativeImageLoader.ALLOWED_FORMATS,randNumGen);
        ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();
        ImageRecordReader recordReader = new ImageRecordReader(height,width,channels,labelMaker);
        recordReader.initialize(data);
        recordReader.setListeners(new LogRecordListener());
        DataSetIterator dataIter = new RecordReaderDataSetIterator(recordReader,batchSize,1,outputNum);
        DataNormalization scaler = new ImagePreProcessingScaler(0,1);
        scaler.fit(dataIter);
        dataIter.setPreProcessor(scaler);
        return dataIter;
    }
```
![model1](https://user-images.githubusercontent.com/15330998/34481737-aae66fa2-efab-11e7-9b89-99fdf29c977f.PNG)

Visualization does not look so good, I am pretty sure I have made some stupid mistake...but no idea what.
Source code says that _outputNum_ variable of my `getData(String type)` method represents the number of possible classes. Since this is a kind of a "one class" situation, should this variable be 5000 (height*width)? I tried changing it to 1 but I got mismatch errors. Sorry to bother with noobie things but I really need to learn this library. Thanks in advance.

**Comment by agibsonccc** [comment]
@Arxa an autoencoder isn't a classifier. That's not really relevant here. Pretrain doesn't actually do anything for you. (This is only used with Variational and normal Autoencoders) You are only using dense layers. Your output layer isn't correct. It should be Activation.IDENTITY . Right now you have a global relu in place. I would suggest maybe trying a different activation function (tanh or sigmoid maybe?) . 

Your learning rate seems a bit high. Drop it down to 1e-3 or lower.

Your setInputType is wrong. You aren't even using cnns. That should be convolutionalFlat .

I would also suggest using a different loss function (RMSE or KL_DIVERGENCE). I'm closing this. Please post questions like this on stack overflow. This is not a github issue. Thanks!
